### ISSUE TITLE
Multiple dataset preparation questions

### ISSUE THREAD

**Issue by jorisbalc** [nanoporetech/remora#66]
Greetings,

I have made a vast amount of progress in my research with remora2.0 but I have a few questions regarding the data needed for the best training results and how one would acquire such form of data.

1. Since remora2.0 the dataset prep step needs an unsorted `.bam` file with the moves table in it. I'm curious to which extent does the basecalling quality matter? We use quite extensive modifications which distort the raw signal from the nanopore and in most cases there are multiple mismatches or deletions around the modified C base (about 2-3 bases to both sides of the modification are called incorrectly, except the modified base which is always called as C). Would qscore filtering matter here? Should it be disabled at all? Would sup models provide any benefits to model training in this manner compared to fast ones?

2. How does remora handle raw reads which could not be mapped in the .bam file? As for example small reads (300-500bps) are impossible to map for me due to the previously mentioned basecalling on modified Cs problem. I've had reads with 60% mismatched bases. However this is slightly less problematic when using longer reads. Is there a way to filter the pod5s by read length before training? I'm sure there's an option in MinKNOW UI during live basecalling, but how would one proceed to filter them further after the sequencing? Edit: I'm very curious now whether there's a simple way to extract only long reads from the pod5s.

3. CG context variety in their surrounding environment goes a long way as I understand. What would you say is the perfect ratio on sequence length and variability and the coverage level? We've had great success with training` GCGC 1` models on a simple 48kb lambda phage genome and ~20m chunks, but not so much while training `CG 0` models. As I understand there must be an adequate coverage count at all times and I'm wondering how this correlates to the lenght and variability of sequence.

4. Does having a larger `can_chunks.npz ` dataset impact the model training in any way and vice versa? Or does remora automatically use similar portions of the datasets in the model training?

Hopefully you are able to clear some of these questions for me and thanks in advance!

Best regards,

JB

**Comment by marcus1487** [comment]
I'll try to touch on the various questions here, but please as specific questions if I missed anything.

> Would qscore filtering matter here? Should it be disabled at all? Would sup models provide any benefits to model training in this manner compared to fast ones?

Filtering input reads is definitely a very important space to explore for modified base detection. We do not have concrete guidelines here as this is still an open are of research, but improving the canonical basecalling model to handle modified bases should certainly be considered for modified bases which distort the basecalls as described here. I would recommend looking at the bonito documentation on basecaller training and raise any issues there with training a bonito model specific to you modified base sample.

> How does remora handle raw reads which could not be mapped in the .bam file? 

Remora currently ignores unmapped reads. Modified base ground truth is linked to reference positions currently in the CLI, so there would be no way to label training sites without the reference. It is possible to construct training sets using the remora API from unmapped reads, but there is not a clear logic as to how sites would be annotated, so no CLI options are available at this time.

> Is there a way to filter the pod5s by read length before training?

Filtering the BAM file should suffice for the purposes of including data in training. Remora will only produce training data from the read ids found in both the POD5 AND the input BAM. Thus filtering your BAM file would be the recommendation here.

> As I understand there must be an adequate coverage count at all times and I'm wondering how this correlates to the lenght and variability of sequence.

More contexts means more for the model to learn and it is thus harder to produce robust models for more contexts. I would suggest that more training data would be advisable for models targeted at more contexts, but we do not have concrete guidelines for this at the moment. 

>Does having a larger can_chunks.npz  dataset impact the model training in any way and vice versa? Or does remora automatically use similar portions of the datasets in the model training?

The next version of remora will have an option to set the labels balance of each batch of training, but this is not currently available. With the current release batches of training are randomly selected from the full training set and thus the label balance will certainly effect the model.

**Comment by jorisbalc** [comment]
Thank you for clearing most of the questions for me,

Regarding the amount of data required to train a model, as mentioned I had a 0,8% average false positive when training GCGC models, but CG models with the same modification are pretty difficult to train for me. The CG models usually results in very poor quality, reaching 30-40% false positives. My question is, is this due to the lambda genome (48kb) being too small or what is the problem here? I would assume GCGC models are way easier to train due to the base context being 4 bases instead of 2. Would using a multiple mB genome increase the quality of the CG model?

Thanks in advance!

**Comment by marcus1487** [comment]
We have not done a large amount of testing on training modified base models from very small genomes, but I would suggest that something much larger than lambda would be recommended. To answer in more detail I would need more information on your training and validation datasets.
